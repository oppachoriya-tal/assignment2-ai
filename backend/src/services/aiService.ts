import { GeminiService } from './geminiService';
import { logger } from '../utils/logger';
import { prisma } from '../config/database';

export interface RecommendationRequest {
  userId: string;
  limit?: number;
  context?: string;
}

export interface RecommendationResponse {
  books: Array<{
    id: string;
    title: string;
    author: string;
    coverImageUrl?: string;
    averageRating: number;
    totalReviews: number;
    reason: string;
    confidence: number;
  }>;
  explanation: string;
}

export interface SimilarBooksRequest {
  bookId: string;
  limit?: number;
}

export interface ReviewAnalysis {
  sentiment: 'positive' | 'negative' | 'neutral';
  themes: string[];
  quality: number; // 0-1 score
  summary: string;
}

export class AIService {
  /**
   * Generate personalized book recommendations using Gemini
   */
  static async generateRecommendations(request: RecommendationRequest): Promise<RecommendationResponse> {
    try {
      return await GeminiService.generateRecommendations(request);
    } catch (error) {
      logger.error('AI recommendation generation failed:', error);
      // Fallback to trending books
      return await this.getTrendingBooksRecommendations(request.limit || 10);
    }
  }

  /**
   * Find similar books using Gemini analysis
   */
  static async findSimilarBooks(request: SimilarBooksRequest): Promise<RecommendationResponse> {
    try {
      return await GeminiService.findSimilarBooks(request);
    } catch (error) {
      logger.error('Similar books generation failed:', error);
      throw error;
    }
  }

  /**
   * Analyze review sentiment and themes using Gemini
   */
  static async analyzeReview(reviewText: string): Promise<ReviewAnalysis> {
    try {
      return await GeminiService.analyzeReview(reviewText);
    } catch (error) {
      logger.error('Review analysis failed:', error);
      return {
        sentiment: 'neutral',
        themes: [],
        quality: 0.5,
        summary: 'Analysis unavailable'
      };
    }
  }

  /**
   * Generate book descriptions using Gemini
   */
  static async generateBookDescription(title: string, author: string, existingDescription?: string): Promise<string> {
    try {
      return await GeminiService.generateText(`Generate a compelling book description for "${title}" by ${author}. ${existingDescription ? `Current description: ${existingDescription}. Please improve it.` : 'Create a new description.'}`);
    } catch (error) {
      logger.error('Book description generation failed:', error);
      return existingDescription || 'Description unavailable';
    }
  }

  /**
   * Get popular books as fallback
   */
  private static async getPopularBooksRecommendations(limit: number): Promise<RecommendationResponse> {
    const books = await prisma.book.findMany({
      take: limit,
      include: {
        reviews: { select: { rating: true } }
      },
      orderBy: {
        reviews: { _count: 'desc' }
      }
    });

    return {
      books: books.map(book => ({
        id: book.id,
        title: book.title,
        author: book.author,
        coverImageUrl: book.coverImageUrl || undefined,
        averageRating: book.reviews.length > 0 
          ? Math.round((book.reviews.reduce((sum, r) => sum + r.rating, 0) / book.reviews.length) * 10) / 10
          : 0,
        totalReviews: book.reviews.length,
        reason: 'Popular among readers',
        confidence: 0.7
      })),
      explanation: 'Popular books recommended for new users'
    };
  }

  /**
   * Get trending books as fallback
   */
  private static async getTrendingBooksRecommendations(limit: number): Promise<RecommendationResponse> {
    const thirtyDaysAgo = new Date();
    thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);

    const books = await prisma.book.findMany({
      where: {
        reviews: {
          some: {
            createdAt: { gte: thirtyDaysAgo }
          }
        }
      },
      take: limit,
      include: {
        reviews: { select: { rating: true } }
      },
      orderBy: {
        reviews: { _count: 'desc' }
      }
    });

    return {
      books: books.map(book => ({
        id: book.id,
        title: book.title,
        author: book.author,
        coverImageUrl: book.coverImageUrl || undefined,
        averageRating: book.reviews.length > 0 
          ? Math.round((book.reviews.reduce((sum, r) => sum + r.rating, 0) / book.reviews.length) * 10) / 10
          : 0,
        totalReviews: book.reviews.length,
        reason: 'Trending recently',
        confidence: 0.8
      })),
      explanation: 'Trending books from the last 30 days'
    };
  }
}
