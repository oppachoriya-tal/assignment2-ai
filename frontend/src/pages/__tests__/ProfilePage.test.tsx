import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import ProfilePage from '../ProfilePage';
import { apiService } from '../../services/api';

// Mock the API service
jest.mock('../../services/api', () => ({
  apiService: {
    makeRequest: jest.fn(),
  },
}));

const mockApiService = apiService as jest.Mocked<typeof apiService>;

// Create mock store
const createMockStore = (isAuthenticated = true) => {
  return configureStore({
    reducer: {
      auth: (state = { isAuthenticated, user: { id: '1', email: 'test@example.com' } }, action) => state,
    },
  });
};

describe('ProfilePage Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders loading state initially', () => {
    mockApiService.makeRequest.mockImplementation(() => new Promise(() => {})); // Never resolves
    
    const store = createMockStore();
    render(
      <Provider store={store}>
        <ProfilePage />
      </Provider>
    );
    
    expect(screen.getByTestId('profile-page')).toBeInTheDocument();
    // Check for skeleton loading elements by class name
    expect(document.querySelectorAll('.MuiSkeleton-root').length).toBeGreaterThan(0);
  });

  it('renders profile data successfully', async () => {
    const mockProfileData = {
      user: {
        id: '1',
        username: 'testuser',
        email: 'test@example.com',
        firstName: 'Test',
        lastName: 'User',
        avatarUrl: null,
        bio: 'Test bio',
        createdAt: '2023-01-01T00:00:00Z',
        lastLogin: '2023-12-01T00:00:00Z'
      },
      reviews: [
        {
          id: 'review-1',
          rating: 5,
          reviewText: 'Great book!',
          createdAt: '2023-11-01T00:00:00Z',
          book: {
            id: 'book-1',
            title: 'Test Book',
            author: 'Test Author',
            coverImageUrl: null
          }
        }
      ],
      totalReviews: 1,
      favorites: [
        {
          id: 'book-1',
          title: 'Test Book',
          author: 'Test Author',
          coverImageUrl: null,
          averageRating: 4.5,
          totalReviews: 10,
          addedToFavorites: '2023-10-01T00:00:00Z'
        }
      ],
      totalFavorites: 1
    };

    mockApiService.makeRequest.mockResolvedValue({
      success: true,
      data: mockProfileData
    });
    
    const store = createMockStore();
    render(
      <Provider store={store}>
        <ProfilePage />
      </Provider>
    );
    
    await waitFor(() => {
      expect(screen.getByText('Test User')).toBeInTheDocument();
      expect(screen.getByText('@testuser')).toBeInTheDocument();
      expect(screen.getByText('test@example.com')).toBeInTheDocument();
      expect(screen.getByText('"Test bio"')).toBeInTheDocument();
    });

    // Check stats cards
    expect(screen.getAllByText('1')[0]).toBeInTheDocument(); // Reviews count
    expect(screen.getByText('Reviews Written')).toBeInTheDocument();
    expect(screen.getByText('Favorite Books')).toBeInTheDocument();
    
    // Check tabs
    expect(screen.getByText('Reviews (1)')).toBeInTheDocument();
    expect(screen.getByText('Favorites (1)')).toBeInTheDocument();
  });

  it('renders error state when API fails', async () => {
    mockApiService.makeRequest.mockRejectedValue(new Error('API Error'));
    
    const store = createMockStore();
    render(
      <Provider store={store}>
        <ProfilePage />
      </Provider>
    );
    
    await waitFor(() => {
      expect(screen.getByText('API Error')).toBeInTheDocument();
    });
  });

  it('renders no data state when API returns no data', async () => {
    mockApiService.makeRequest.mockResolvedValue({
      success: false,
      message: 'Failed to load profile data'
    });
    
    const store = createMockStore();
    render(
      <Provider store={store}>
        <ProfilePage />
      </Provider>
    );
    
    await waitFor(() => {
      expect(screen.getByText('Failed to load profile data')).toBeInTheDocument();
    });
  });

  it('switches between tabs correctly', async () => {
    const mockProfileData = {
      user: {
        id: '1',
        username: 'testuser',
        email: 'test@example.com',
        firstName: 'Test',
        lastName: 'User',
        avatarUrl: null,
        bio: 'Test bio',
        createdAt: '2023-01-01T00:00:00Z',
        lastLogin: '2023-12-01T00:00:00Z'
      },
      reviews: [],
      totalReviews: 0,
      favorites: [],
      totalFavorites: 0
    };

    mockApiService.makeRequest.mockResolvedValue({
      success: true,
      data: mockProfileData
    });
    
    const store = createMockStore();
    render(
      <Provider store={store}>
        <ProfilePage />
      </Provider>
    );
    
    await waitFor(() => {
      expect(screen.getByText('Test User')).toBeInTheDocument();
    });

    // Check that Reviews tab is active by default
    expect(screen.getByText('No reviews yet')).toBeInTheDocument();
    expect(screen.getByText('Start reviewing books to see them here!')).toBeInTheDocument();
  });

  it('has correct container structure', async () => {
    const mockProfileData = {
      user: {
        id: '1',
        username: 'testuser',
        email: 'test@example.com',
        firstName: 'Test',
        lastName: 'User',
        avatarUrl: null,
        bio: 'Test bio',
        createdAt: '2023-01-01T00:00:00Z',
        lastLogin: '2023-12-01T00:00:00Z'
      },
      reviews: [],
      totalReviews: 0,
      favorites: [],
      totalFavorites: 0
    };

    mockApiService.makeRequest.mockResolvedValue({
      success: true,
      data: mockProfileData
    });
    
    const store = createMockStore();
    const { container } = render(
      <Provider store={store}>
        <ProfilePage />
      </Provider>
    );
    
    await waitFor(() => {
      expect(screen.getByText('Test User')).toBeInTheDocument();
    });

    const profileContainer = screen.getByTestId('profile-page');
    expect(profileContainer).toBeInTheDocument();
    expect(profileContainer).toHaveClass('MuiContainer-root');
  });

  it('renders consistently', async () => {
    const mockProfileData = {
      user: {
        id: '1',
        username: 'testuser',
        email: 'test@example.com',
        firstName: 'Test',
        lastName: 'User',
        avatarUrl: null,
        bio: 'Test bio',
        createdAt: '2023-01-01T00:00:00Z',
        lastLogin: '2023-12-01T00:00:00Z'
      },
      reviews: [],
      totalReviews: 0,
      favorites: [],
      totalFavorites: 0
    };

    mockApiService.makeRequest.mockResolvedValue({
      success: true,
      data: mockProfileData
    });
    
    const store = createMockStore();
    const { container } = render(
      <Provider store={store}>
        <ProfilePage />
      </Provider>
    );
    
    await waitFor(() => {
      expect(screen.getByText('Test User')).toBeInTheDocument();
    });
    
    expect(container.firstChild).toMatchSnapshot();
  });
});