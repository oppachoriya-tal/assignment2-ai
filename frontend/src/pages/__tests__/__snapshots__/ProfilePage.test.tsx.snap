// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage Component renders consistently 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  data-testid="profile-page"
>
  <div
    class="MuiBox-root css-8f57v9"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-13nu2j-MuiPaper-root-MuiCard-root"
      style="opacity: 1; webkit-transition: opacity 800ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 800ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiBox-root css-qp976a"
      />
      <div
        class="MuiCardContent-root css-rt41h5-MuiCardContent-root"
      >
        <div
          class="MuiStack-root css-6xevoz-MuiStack-root"
        >
          <span
            class="MuiBadge-root css-1c32n2y-MuiBadge-root"
            style="transform: none; webkit-transition: transform 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 1000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1y5adst-MuiAvatar-root"
            >
              T
              U
            </div>
            <span
              class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-9sufak-MuiBadge-badge"
            >
              <div
                class="MuiBox-root css-blyjw4"
              >
                <span
                  class="MuiTypography-root MuiTypography-caption css-hlcx9g-MuiTypography-root"
                >
                  âœ“
                </span>
              </div>
            </span>
          </span>
          <div
            class="MuiBox-root css-1rr4qq7"
          >
            <h1
              class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-1h9kp1f-MuiTypography-root"
            >
              Test
               
              User
            </h1>
            <div
              class="MuiStack-root css-1irh47u-MuiStack-root"
            >
              <div
                class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-9q0yi2-MuiChip-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconMedium MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
                  />
                </svg>
                <span
                  class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                >
                  @testuser
                </span>
              </div>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1oyxbyq-MuiChip-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
                  data-testid="LanguageIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2m6.93 6h-2.95c-.32-1.25-.78-2.45-1.38-3.56 1.84.63 3.37 1.91 4.33 3.56M12 4.04c.83 1.2 1.48 2.53 1.91 3.96h-3.82c.43-1.43 1.08-2.76 1.91-3.96M4.26 14C4.1 13.36 4 12.69 4 12s.1-1.36.26-2h3.38c-.08.66-.14 1.32-.14 2 0 .68.06 1.34.14 2zm.82 2h2.95c.32 1.25.78 2.45 1.38 3.56-1.84-.63-3.37-1.9-4.33-3.56m2.95-8H5.08c.96-1.66 2.49-2.93 4.33-3.56C8.81 5.55 8.35 6.75 8.03 8M12 19.96c-.83-1.2-1.48-2.53-1.91-3.96h3.82c-.43 1.43-1.08 2.76-1.91 3.96M14.34 14H9.66c-.09-.66-.16-1.32-.16-2 0-.68.07-1.35.16-2h4.68c.09.65.16 1.32.16 2 0 .68-.07 1.34-.16 2m.25 5.56c.6-1.11 1.06-2.31 1.38-3.56h2.95c-.96 1.65-2.49 2.93-4.33 3.56M16.36 14c.08-.66.14-1.32.14-2 0-.68-.06-1.34-.14-2h3.38c.16.64.26 1.31.26 2s-.1 1.36-.26 2z"
                  />
                </svg>
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  English
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1mo5v2s-MuiTypography-root"
            >
              test@example.com
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-us5c47-MuiTypography-root"
            >
              "
              Test bio
              "
            </p>
            <div
              class="MuiStack-root css-p81m6m-MuiStack-root"
            >
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1oyxbyq-MuiChip-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
                  data-testid="CalendarTodayIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 18H4V8h16z"
                  />
                </svg>
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  Joined January 1, 2023
                </span>
              </div>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1oyxbyq-MuiChip-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5M12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5m0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"
                  />
                </svg>
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  Last active December 1, 2023
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiStack-root css-jfdv4h-MuiStack-root"
          >
            <button
              aria-label="Edit Profile"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1x3cudu-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                />
              </svg>
            </button>
            <button
              aria-label="Share Profile"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1x3cudu-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ShareIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-1ov6x74-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-3 css-10voxkt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-7zoerl-MuiPaper-root-MuiCard-root"
          style="transform: none; webkit-transition: transform 1200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 1200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <div
            class="MuiBox-root css-1hlpz3m"
          />
          <div
            class="MuiCardContent-root css-993mr5-MuiCardContent-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1t212i8-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
            <div
              class="MuiTypography-root MuiTypography-h3 css-1sgmnm5-MuiTypography-root"
            >
              0
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-ncm894-MuiTypography-root"
            >
              Reviews Written
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-3 css-10voxkt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-86hvxe-MuiPaper-root-MuiCard-root"
          style="transform: none; webkit-transition: transform 1400ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 1400ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <div
            class="MuiBox-root css-1hlpz3m"
          />
          <div
            class="MuiCardContent-root css-993mr5-MuiCardContent-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1t212i8-MuiSvgIcon-root"
              data-testid="FavoriteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
              />
            </svg>
            <div
              class="MuiTypography-root MuiTypography-h3 css-1sgmnm5-MuiTypography-root"
            >
              0
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-ncm894-MuiTypography-root"
            >
              Favorite Books
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-3 css-10voxkt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-hmzfc6-MuiPaper-root-MuiCard-root"
          style="transform: none; webkit-transition: transform 1600ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 1600ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <div
            class="MuiBox-root css-1hlpz3m"
          />
          <div
            class="MuiCardContent-root css-993mr5-MuiCardContent-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1t212i8-MuiSvgIcon-root"
              data-testid="TrendingUpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m16 6 2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
              />
            </svg>
            <div
              class="MuiTypography-root MuiTypography-h3 css-1sgmnm5-MuiTypography-root"
            >
              0.0
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-ncm894-MuiTypography-root"
            >
              Avg Rating Given
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-3 css-10voxkt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-jjy6xt-MuiPaper-root-MuiCard-root"
          style="transform: none; webkit-transition: transform 1800ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 1800ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <div
            class="MuiBox-root css-1hlpz3m"
          />
          <div
            class="MuiCardContent-root css-993mr5-MuiCardContent-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1t212i8-MuiSvgIcon-root"
              data-testid="SchoolIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 13.18v4L12 21l7-3.82v-4L12 17zM12 3 1 9l11 6 9-4.91V17h2V9z"
              />
            </svg>
            <div
              class="MuiTypography-root MuiTypography-h3 css-1sgmnm5-MuiTypography-root"
            >
              USER
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-ncm894-MuiTypography-root"
            >
              Member Since
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-169956t-MuiPaper-root"
      style="opacity: 1; webkit-transition: opacity 2000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 2000ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiTabs-root css-80by1j-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            aria-label="profile tabs"
            class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
            role="tablist"
          >
            <button
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiStack-root css-kdc3n8-MuiStack-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CommentIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4zM18 14H6v-2h12zm0-3H6V9h12zm0-3H6V6h12z"
                  />
                </svg>
                <span>
                  Reviews (
                  0
                  )
                </span>
              </div>
              <span
                class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              />
            </button>
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <div
                class="MuiStack-root css-kdc3n8-MuiStack-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="FavoriteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
                  />
                </svg>
                <span>
                  Favorites (
                  0
                  )
                </span>
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-1wcaknn"
      >
        <div
          class="MuiBox-root css-0"
          style="opacity: 1; webkit-transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <div
            class="MuiStack-root css-1dipck7-MuiStack-root"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1xxajqc-MuiTypography-root"
            >
              Recent Reviews
            </h5>
            <div
              class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-outlinedPrimary css-1y6wsfa-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
              >
                0 total
              </span>
            </div>
          </div>
          <div
            class="MuiBox-root css-fvfjnh"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11ndtqy-MuiSvgIcon-root"
              data-testid="BookIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 2H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2M6 4h5v8l-2.5-1.5L6 12z"
              />
            </svg>
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-137rogq-MuiTypography-root"
            >
              No reviews yet
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              Start reviewing books to see them here!
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
